external authorization: set the SNI value from server name if it isn't available on the connection/socket

marc-barry opened the pull request:
This is my first commit to the Envoy project and I haven't written C++ in many years. I'm still navigating the types and hierarchy and best practices for obtaining the data I need to complete this pull request. I have started in `draft` first to see the CI processes and look at how the tests run and if my changes cause any issues with current tests.

<!--
!!!ATTENTION!!!

If you are fixing *any* crash or *any* potential security issue, *do not*
open a pull request in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.
Thank you in advance for helping to keep Envoy secure.

!!!ATTENTION!!!

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)
-->

Commit Message: external authorization: set the SNI value from server name if it isn't available on the connection/socket
Additional Description: Leverages the TLS inspectors server name value, if one was set.
Risk Level: low
Testing: Will test that the SNI value is set from the server name of a connection when the TLS session doesn't have the SNI.
Docs Changes: N/A
Release Notes: exterbal authorization
Platform Specific Features: N/A
[Optional Runtime guard:]
Fixes https://github.com/envoyproxy/envoy/issues/34002
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

marc-barry commented:
I'm unable to determine which of the unit tests in `envoy/test/extensions/filters/common/ext_authz/check_request_utils_test.cc` is failing. I know that is the location that CI reports to me but I don't know the Envoy CI system well enough to know how to determine which tests is the issue.

fredyw commented:
> I'm unable to determine which of the unit tests in `envoy/test/extensions/filters/common/ext_authz/check_request_utils_test.cc` is failing. I know that is the location that CI reports to me but I don't know the Envoy CI system well enough to know how to determine which tests is the issue.

A passing by comment:

In the `envoy-presubmit` Azure Pipelines, click on [View more details on Azure Pipelines](https://dev.azure.com/cncf/4684fb3d-0389-4e0b-8251-221942316e06/_build/results?buildId=170326) at the bottom. That'll bring you to the Azure Devops Pipelines and simply click on the error.

This is the actual error: https://dev.azure.com/cncf/envoy/_build/results?buildId=170326&view=logs&j=8c169225-0ae8-53bd-947f-07cb81846cb5&t=d1a98671-b7ba-5fbf-f06c-ff337c010df4&l=249

ggreenway commented on source/extensions/filters/common/ext_authz/check_request_utils.cc:
```diff
@@ -248,8 +250,14 @@ void CheckRequestUtils::createHttpCheck(
+  if (include_tls_session) {
+    // Try to get the SNI from the TLS session. If not available there then try the
+    // data from the TLS inspector (i.e. the server name)
+    if (cb->connection()->ssl() != nullptr) {
```
To avoid the copy/paste, I'd change `setTLSSession()` to take `cb->connection()` as the second parameter, and put the logic of `connection->ssl() != nullptr` in that function, and then get `connection->requestedServerName()` as the alternative in that function. Also, if `requestedServerName().empty()` is true, don't make the call to `set_sni()` because `mutable_tls_session()` would then have the side effect of creating an empty message; no message at all is the correct behavior in this case.

marc-barry replied:
I made this change. Thanks for the recommendation.

ggreenway reviewed (changes requested):
Thanks for working on this!

Please add a test case for this new logic, and you'll need to add a release note.

/wait

ggreenway commented on source/extensions/filters/common/ext_authz/check_request_utils.h:
```diff
@@ -140,7 +140,7 @@ class CheckRequestUtils {
       bool encode_raw_headers, const MatcherSharedPtr& request_header_matchers);
   static void setTLSSession(envoy::service::auth::v3::AttributeContext::TLSSession& session,
-                            const Ssl::ConnectionInfoConstSharedPtr ssl_info);
+                            OptRef<const Envoy::Network::Connection> connection);
```
```suggestion
                            const Envoy::Network::Connection& connection);
```

It's not an optional parameter; it always has a value. I realize that `StreamFilterCallbacks::connection()` returns an OptRef, but it's already assumed to have a value and dereferenced in other places in `createHttpCheck()`, so just do the same for calling this function.

marc-barry replied:
https://github.com/envoyproxy/envoy/pull/34100/commits/a49951c222c53f3ef2cde3a1b9843582ca9b39d8

ggreenway commented on source/extensions/filters/common/ext_authz/check_request_utils.cc:
```diff
@@ -214,12 +214,17 @@ void CheckRequestUtils::setAttrContextRequest(
-  if (!ssl_info->sni().empty()) {
-    const std::string server_name(ssl_info->sni());
-    session.set_sni(server_name);
+void setTLSSession(envoy::service::auth::v3::AttributeContext::TLSSession& session,
+                   OptRef<const Envoy::Network::Connection> connection) {
+  if (connection.has_value()) {
+    const Ssl::ConnectionInfoConstSharedPtr ssl_info = connection->ssl();
+    if (ssl_info != nullptr && !ssl_info->sni().empty()) {
+      const std::string server_name(ssl_info->sni());
+      session.set_sni(server_name);
+    } else if (!connection->requestedServerName().empty()) {
+      const std::string server_name(connection->requestedServerName());
+      session.set_sni(server_name);
+    }
```
```suggestion
void setTLSSession(envoy::service::auth::v3::AttributeContext::TLSSession& session,
                   const Network::Connection& connection) {
    const Ssl::ConnectionInfoConstSharedPtr ssl_info = connection.ssl();
    if (ssl_info != nullptr && !ssl_info->sni().empty()) {
      const std::string server_name(ssl_info->sni());
      session.set_sni(server_name);
    } else if (!connection.requestedServerName().empty()) {
      const std::string server_name(connection.requestedServerName());
      session.set_sni(server_name);
```

(This isn't correctly formatted; don't accept it directly. Take it and run clang-format.)

marc-barry replied:
https://github.com/envoyproxy/envoy/pull/34100/commits/a49951c222c53f3ef2cde3a1b9843582ca9b39d8

ggreenway reviewed (changes requested):
Please add some test coverage for the new functionality.

/wait

marc-barry commented:
I'm working on tests now. Both for the new functionality and addressing the tests that this new functionality is causing to fail.

marc-barry commented:
@ggreenway I addressed all you comments. I added a new unit test for TCP and then adapted `callHttpCheckAndValidateRequestAttributes`, which is called by the HTTP tests, to conditionally perform a check using the correct source of the SNI value for comparison. I also adapted the existing `CheckAttrContextPeerTLSSession` test to have some additional checks that certain functions are called a certain number of times, such as `requestedServerName()`.

ggreenway reviewed (changes requested):
Looks good! 

Please add a release note to `changelogs/current.yaml`.

/wait

marc-barry commented:
@ggreenway is there anything I need to do for that failing check? It does look to me that it might be CI flakiness but perhaps there is an underlying issue that needs my attention.

ggreenway commented:
That's a weird error; pretty sure you didn't cause it. I'm re-running the failing job to see if it passes.
